### PR TITLE
DRAFT: Attempt to port doctr to generator rendering

### PR DESCRIPTION
**Pull request by mara004**
(See https://github.com/mindee/doctr/discussions/1000 for background)

A problem with generators was missing `__len__()`. However, we know the prospected length already, so we can create a wrapper object that behaves like an iterator but also implements `__len__()`.

~~Backwards-compatible~~ pypdfium2 4 port on the way.

### CODE REVIEW

**Comment by mara004** [comment]
Ah, the CI results show that quite some more changes may be necessary (I only ran `tests/common` locally, so didn't notice).

**Comment by mara004** [comment]
Looks like CI has halted itself?

**Comment by mara004** [comment]
So far, among the 6 failing checks, only the style failures seem valid. The others are (I think) unrelated to my changes, but 10 checks are still in progress.

**Comment by mara004** [comment]
Not sure if the remaining checks will ever finish? Some steps are running since >5h already...

File: scripts/analyze.py [hunk 1]
```diff
@@ -26,7 +26,8 @@ def main(args):
     model = ocr_predictor(args.detection, args.recognition, pretrained=True)
 
     if args.path.lower().endswith(".pdf"):
-        doc = DocumentFile.from_pdf(args.path)
+        # FIXME avoid casting to list
```
[the remark this answers is not in the document]
Comment (mara004) [reply scripts/analyze.py]: black complains about the space change, but I still have to apply the black/isort/pydocstyle suggestions anyway.

**Comment by mara004** [comment]
I've just merged the devel branch, so pypdfium2 v4 will be released soon (probably next monday).
This means at least part of the patch should be interesting.

File: doctr/io/pdf.py [hunk 2]
```diff
@@ -3,44 +3,68 @@
...
+        pdf = pdfium.PdfDocument(file, password=password)
+
+        if page_indices:
+            self._len = len(page_indices)
+        else:
+            self._len = len(pdf)
+
+        render_kwargs = dict(scale=scale, page_indices=page_indices, rev_byteorder=rgb_mode, **kwargs)
+        if hasattr(pdf, "render_to"):  # v3 compat
+            self._generator = (p for p, _ in pdf.render_to(pdfium.BitmapConv.numpy_ndarray, **render_kwargs))
```
Comment (mara004) [inline doctr/io/pdf.py:59]: ... or just drop v3. v4 is nicer, and no relevant dependency changes, so technically no point messing up the code with backwards compatibility

File: doctr/io/pdf.py [hunk 3]
```diff
@@ -3,44 +3,62 @@
...
+        return self._len
+
+    def __next__(self) -> np.ndarray:
+        self._len -= 1
+        return next(self._generator)
+
+    def __iter__(self) -> Generator[np.ndarray, None, None]:
+        for page in self._generator:
+            self._len -= 1
+            yield page
```
Comment (mara004) [inline doctr/io/pdf.py:64]: Better alternative
```python
while True:
    try:
        yield next(self)
    except StopIteration:
        break
```

This avoids duplication if further work needs to be done on each step.

**Comment by felixdittrich92** [comment]
Hi @mara004 :wave: 
Thanks for the PR :)
I'm not sure if the changes will add any real value/improvement as long as we end up being forced to cast to a list. 
(Otherwise we would have to refactor all the logic for processing PDF files).
Also ping @frgfm  for a second opinion.

File: doctr/io/pdf.py (hunk 3 above)
Comment (felixdittrich92) [reply doctr/io/pdf.py:64]: Looks better :+1:

**Comment by mara004** [comment]
Hmm, yeah, I see the problem. However, it might be a step in the right direction, anyway?
At least it would be an advantage for callers using `DocumentFile.from_pdf()` with their own code.

**Comment by mara004** [comment]
Just in case you intend to merge this, please squash the commits first. Thanks!

**Comment by mara004** [comment]
Another thing: pypdfium2 doesn't permit rendering in parallel with buffer/bytes input anymore due to unresolved random segfaults. The tests expect this to work, though. Should we handle these input types by saving to a tempfile in `PdfReader`?

**Comment by frgfm** [comment]
Hey everyone :wave:

My apologies, I didn't come back to this notification before now! Glad to see you around @mara004 :smile: 
A few high-level notes:
- I fully agree with the proposal of moving list to generator (that will save RAM on the end device and I think we could make it compatible with the overall lib with a bit of planning)
- interface-wise, moving from functional to class version is alright in my opinion

However, I suggest that if we make those changes we check if harmonizing this would make sense. I mean:
- for other inputs than PDFs (images, etc.), does it make sense to have a generator
- functional vs class API for the whole io module
The few challenges I imagine: generators are actually helpful if the inputs are file paths (otherwise it's the same for the RAM), and we should check performance impact on the transformation/preprocessing stage.

What do you think?

**Comment by mara004** [comment]
Thanks for the response and the interesting considerations! A few comments/clarifications:
* I'm not sure how doctr handles images, but as I guess compressed inputs (e. g.  JPEG, PNG) need to be decompressed for processing, I imagine doing that in step-by-step fashion would be good, too.
* File paths/handles are certainly more memory efficient than a bytes object. However, I believe delaying rendering/decompression is a logically independent memory advantage.

File: scripts/analyze.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply scripts/analyze.py]: A bit late but you can do this by simply running `make style` and check typings, etc. with `make quality` :)

[the remark this answers is not in the document]
Comment (mara004) [reply scripts/analyze.py]: Thanks!

**Comment by felixdittrich92** [comment]
hi @mara004 :wave:, 

for the moment i would suggest to do only minimal changes without breaking our IO api to be compatible with your latest releases >=4.0.0. 

If have tested it shortly:

```
def read_pdf(
    file: AbstractFile,
    scale: float = 2,
    rgb_mode: bool = True,
    password: Optional[str] = None,
    **kwargs: Any,
) -> List[np.ndarray]:
    """Read a PDF file and convert it into an image in numpy format

    >>> from doctr.documents import read_pdf
    >>> doc = read_pdf("path/to/your/doc.pdf")

    Args:
        file: the path to the PDF file
        scale: rendering scale (1 corresponds to 72dpi)
        rgb_mode: if True, the output will be RGB, otherwise BGR
        password: a password to unlock the document, if encrypted
        kwargs: additional parameters to :meth:`pypdfium2.PdfDocument.render`

    Returns:
        the list of pages decoded as numpy ndarray of shape H x W x C
    """
    
    # Rasterise pages to numpy ndarrays with pypdfium2
    pdf = pdfium.PdfDocument(file, password=password, autoclose=True)
    renderer = pdf.render(pdfium.PdfBitmap.to_numpy, scale=scale, rev_byteorder=rgb_mode, **kwargs)
    return [img for img in renderer]
```

but it blocks all which is not a Path or str:

```
if not isinstance(self._input, (Path, str)):
     raise ValueError("Can only render in parallel with file path input.")
```

**Comment by mara004** [comment]
Sure, you can do that. (However, I would still appreciate if we can finish the generator work some day. :) )

> but it blocks all which is not a Path or str:

Yes, unfortunately I had to make that change due to random segfaults when rendering in parallel with bytes/buffers (https://github.com/pypdfium2-team/pypdfium2/issues/164). This is strange because it works perfectly without multiprocessing, and I'm afraid I was unable to locate the cause so far. These failures only appeared after I updated my Python/OS, on the previous install and GH runners the test suite had always passed...

**Comment by felixdittrich92** [comment]
> Sure, you can do that. (However, I would still appreciate if we can finish the generator work some day. :) )
> 
> > but it blocks all which is not a Path or str:
> 
> Yes, unfortunately I had to make that change due to random segfaults when rendering in parallel with bytes/buffers ([pypdfium2-team/pypdfium2#164](https://github.com/pypdfium2-team/pypdfium2/issues/164)). This is strange because it works perfectly without multiprocessing, and I'm afraid I was unable to locate the cause so far. These failures only appeared after I updated my Python/OS, on the previous install and GH runners the test suite had always passed...

Sure i will create an issue to keep it on track :+1: 

Mh the problem is also if i try to deactivate the parallel rendering with `n_processes=1` it has no effect which is currently a blocker because it seems there is no way to work with `bytes` inside `render` ? :sweat_smile:

**Comment by mara004** [comment]
> Mh the problem is also if i try to deactivate the parallel rendering with n_processes=1 it has no effect which is currently a blocker because it seems there is no way to work with bytes inside render ? 😅

Well, you can use the page-level renderer which is not parallelized, something like `[page.render(...) for page in pdf]`.
However, that is not recommendable. I would write a tempfile instead, will be much faster than bytes without concurrency.

Also previously, where (unsafe) parallel bytes rendering was possible, I doubt if it was good since you then have the pdf in memory `n_processes+1` times, which can be suboptimal for big pdfs.

**Pull request closed without merging**